src/winkerberos.c

```diff
@@ -437,8 +437,136 @@ sspi_client_clean(PyObject* self, PyObject* args) {
+        "acceptor_address", "application_data", NULL};
+    SEC_CHANNEL_BINDINGS* channel_bindings;
+    SecPkgContext_Bindings* context_bindings;
+    PyObject *pychan_bindings = NULL;
+    int result = 0;
+    int initiator_addrtype = 0;
+    int acceptor_addrtype = 0;
+    int offset = 0;
+    int data_length = 0;
+
+    int *initiator_length = 0;
+    int *acceptor_length = 0;
+    int *application_length = 0;
+
+    unsigned char* offset_p = NULL;
+
+    const char *encoding = NULL;
+    char **initiator_address = NULL;
+    char **acceptor_address = NULL;
+    char **application_data = NULL;
+
+    if (!PyArg_ParseTupleAndKeywords(args, keywds, "|iet#iet#et#", kwlist,
+            &initiator_addrtype, &encoding, &initiator_address, &initiator_length,
+            &acceptor_addrtype, &encoding, &acceptor_address, &acceptor_length,
+            &encoding, &application_data, &application_length)) {
+        return NULL;
+    }
+
+    data_length = (int)initiator_length + (int)acceptor_length + (int)application_length;
+    offset = sizeof(SEC_CHANNEL_BINDINGS);
+
+    context_bindings = (SecPkgContext_Bindings *) malloc(sizeof(SecPkgContext_Bindings));
+    context_bindings->BindingsLength = sizeof(SEC_CHANNEL_BINDINGS) + data_length;
+
+    channel_bindings = (SEC_CHANNEL_BINDINGS*) malloc(context_bindings->BindingsLength);
+    context_bindings->Bindings = channel_bindings;
+
+    channel_bindings->dwInitiatorAddrType = initiator_addrtype;
+    channel_bindings->cbInitiatorLength = initiator_length;
+    if (initiator_address != NULL) {
+        channel_bindings->dwInitiatorOffset = offset;
+
+        offset_p = &((unsigned char*) channel_bindings)[offset];
+        memcpy(&offset_p[0], initiator_address, initiator_length);
+        PyMem_Free(initiator_address);
+        offset = offset + (int)initiator_length;
+    } else {
+        channel_bindings->dwInitiatorOffset = 0;
+    }
+
+    channel_bindings->dwAcceptorAddrType = acceptor_addrtype;
+    channel_bindings->cbAcceptorLength = acceptor_length;
+    if (acceptor_address != NULL) {
+        channel_bindings->dwAcceptorOffset = offset;
+
+        offset_p = &((unsigned char*) channel_bindings)[offset];
+        memcpy(&offset_p[0], acceptor_address, acceptor_length);
+        PyMem_Free(acceptor_address);
+        offset = offset + (int)acceptor_length;
+    } else {
+        channel_bindings->dwAcceptorOffset = 0;
+    }
+
+    channel_bindings->cbApplicationDataLength = application_length;
+    if (application_data != NULL) {
+        channel_bindings->dwApplicationDataOffset = offset;
+        offset_p = &((unsigned char*) channel_bindings)[offset];
+        memcpy(&offset_p[0], application_data, application_length);
+        PyMem_Free(application_data);
+    } else {
+        channel_bindings->dwApplicationDataOffset = 0;
+    }
+
+    pychan_bindings = PyCObject_FromVoidPtr(context_bindings, &destruct_channel_bindings_struct);
```

Comment:
Note that if this fails you will have to free channel_bindings and context_bindings before returning NULL.